File: mopro_flutter_plugin/ios/Classes/MoproFlutterPlugin.swift
import Flutter
import moproFFI
import UIKit

public class MoproFlutterPlugin: NSObject, FlutterPlugin {
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "mopro_flutter", binaryMessenger: registrar.messenger())
        let instance = MoproFlutterPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "generateProof":
            guard let args = call.arguments as? [String: Any],
                  let zkeyPath = args["zkeyPath"] as? String,
                  let inputs = args["inputs"] as? [String: [String]]
            else {
                result(FlutterError(code: "ARGUMENT_ERROR", message: "Missing arguments", details: nil))
                return
            }

            do {
                // Call the function from mopro.swift
                let proofResult = try generateCircomProof(zkeyPath: zkeyPath, circuitInputs: inputs)
                let proof = toEthereumProof(proof: proofResult.proof)
                let convertedInputs = toEthereumInputs(inputs: proofResult.inputs)
               
                let proofList: [[String: Any]] = [
                    ["x": proof.a.x, "y": proof.a.y],
                    ["x": proof.b.x, "y": proof.b.y],
                    ["x": proof.c.x, "y": proof.c.y]
                ]
                
                let resMap: [String: Any] = [
                    "proof": proofList,
                    "inputs": convertedInputs
                ]
                
                // Return the proof and inputs as a map supported by the StandardMethodCodec
                result(resMap)
            } catch {
                result(FlutterError(code: "PROOF_GENERATION_ERROR", message: "Failed to generate proof", details: error.localizedDescription))
            }

        default:
            result(FlutterMethodNotImplemented)
        }
    }
}

File: mopro_flutter_plugin/android/src/main/kotlin/com/example/mopro_flutter/MoproFlutterPlugin.kt
package com.example.mopro_flutter

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

import uniffi.mopro.generateCircomProof

import io.flutter.plugin.common.StandardMethodCodec
import uniffi.mopro.ProofCalldata
import uniffi.mopro.toEthereumInputs
import uniffi.mopro.toEthereumProof
import uniffi.mopro.ProofLib
/** MoproFlutterPlugin */
class MoproFlutterPlugin : FlutterPlugin, MethodCallHandler {
    /// The MethodChannel that will the communication between Flutter and native Android
    ///
    /// This local reference serves to register the plugin with the Flutter Engine and unregister it
    /// when the Flutter Engine is detached from the Activity
    private lateinit var channel: MethodChannel

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(
            flutterPluginBinding.binaryMessenger,
            "mopro_flutter",
            StandardMethodCodec.INSTANCE
        )
        channel.setMethodCallHandler(this)
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        if (call.method == "generateProof") {
            val zkeyPath = call.argument<String>("zkeyPath") ?: return result.error(
                "ARGUMENT_ERROR",
                "Missing zkeyPath",
                null
            )

            val inputs =
                call.argument<String>("inputs") ?: return result.error(
                    "ARGUMENT_ERROR",
                    "Missing inputs",
                    null
                )

            val res = generateCircomProof(zkeyPath, inputs, ProofLib.ARKWORKS)
            val proof: ProofCalldata = toEthereumProof(res.proof)
            val convertedInputs: List<String> = toEthereumInputs(res.inputs)

            val proofList = listOf(
                mapOf(
                    "x" to proof.a.x,
                    "y" to proof.a.y
                ), mapOf(
                    "x" to proof.b.x,
                    "y" to proof.b.y
                ), mapOf(
                    "x" to proof.c.x,
                    "y" to proof.c.y
                )
            )

            // Return the proof and inputs as a map supported by the StandardMethodCodec
            val resMap = mapOf(
                "proof" to proofList,
                "inputs" to convertedInputs
            )

            result.success(
                resMap
            )
        } else {
            result.notImplemented()
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }
}
